futures-util/src/stream/diff_ascending.rs

```diff
@@ -0,0 +1,123 @@
+use std::cmp::Ordering;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use Poll::*;
+
+use crate::stream::{Fuse, StreamExt};
+use futures_core::stream::{FusedStream, Stream};
+use pin_project_lite::pin_project;
+
+/// Diff two sorted streams in constant space. `diff_ascending(x, y)` returns a
+/// stream containing all elements in `x` not present in `y`. This operation is
+/// not commutative. The precondition (input sterams are ascending) is not
+/// checked.
+///
+/// ```
+/// # futures::executor::block_on(async {
+/// use futures::stream::{self, StreamExt};
+///
+/// let s1 = stream::iter(0..10);
+/// let s2 = stream::iter(0..5);
+/// let collected: Vec<i32> = stream::diff_ascending(s1, s2).collect().await;
+/// assert_eq!(collected, vec![5,6,7,8,9]);
+/// # });
+/// ```
+pub fn diff_ascending<T: Ord, St1: Stream<Item = T>, St2: Stream<Item = T>>(
+    st1: St1,
+    st2: St2,
+) -> DiffAscending<T, St1, St2> {
+    DiffAscending {
+        left: st1.fuse(),
+        right: st2.fuse(),
+        left_peek: None,
+        right_peek: None,
+        right_terminated: false,
+    }
+}
+
+pin_project! {
+    /// Struct for the [`diff_ascending`] method.
+    #[derive(Debug)]
+    #[must_use = "streams do nothing unless polled"]
+    pub struct DiffAscending<T, St1: Stream<Item = T>, St2: Stream<Item = T>> {
+        #[pin]
+        left: Fuse<St1>,
+        #[pin]
+        right: Fuse<St2>,
+        left_peek: Option<T>,
+        right_peek: Option<T>,
+        right_terminated: bool,
+    }
+}
+
+impl<T: Ord, St1: Stream<Item = T>, St2: Stream<Item = T>> Stream for DiffAscending<T, St1, St2> {
+    type Item = T;
+
+    // For reference
+    // diff :: Ord a => [a] -> [a] -> [a]
+    // diff (x : xs) (y : ys)
+    //   | x == y = diff xs ys
+    //   | x < y = x : diff xs (y : ys)
+    //   | x > y = diff (x : xs) ys
+    // diff xs [] = xs
+    // diff [] _ = []
+
```

Comment:
I prototyped the actual algorithm in Haskell. Happy to delete this if it's not helpful.